homeassistant/components/azure_data_explorer/config_flow.py

```diff
@@ -0,0 +1,155 @@
+"""Config flow for Azure Data Explorer integration."""
+from __future__ import annotations
+
+from copy import deepcopy
+import logging
+from typing import Any
+
+from azure.kusto.data.exceptions import KustoAuthenticationError, KustoServiceError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+
+from . import AzureDataExplorerClient
+from .const import (
+    CONF_ADX_CLUSTER_INGEST_URI,
+    CONF_ADX_DATABASE_NAME,
+    CONF_ADX_TABLE_NAME,
+    CONF_APP_REG_ID,
+    CONF_APP_REG_SECRET,
+    CONF_AUTHORITY_ID,
+    CONF_SEND_INTERVAL,
+    CONF_USE_FREE,
+    DEFAULT_OPTIONS,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ADX_CLUSTER_INGEST_URI): str,
+        vol.Required(CONF_ADX_DATABASE_NAME): str,
+        vol.Required(CONF_ADX_TABLE_NAME): str,
+        vol.Required(CONF_APP_REG_ID): str,
+        vol.Required(CONF_APP_REG_SECRET): str,
+        vol.Required(CONF_AUTHORITY_ID): str,
+        vol.Optional(CONF_USE_FREE, default=False): bool,
+    }
+)
+
+
+async def validate_input(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> dict[str, Any] | None:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    client = AzureDataExplorerClient(
+        clusteringesturi=data["clusteringesturi"],
+        database=data["database"],
+        table=data["table"],
+        client_id=data["client_id"],
+        client_secret=data["client_secret"],
+        authority_id=data["authority_id"],
+        use_free_cluster=data["use_free_cluster"],
+    )
+
+    try:
+        await hass.async_add_executor_job(client.test_connection)
+
+    except KustoAuthenticationError as exp:
+        _LOGGER.error(exp)
+        return {"base": "invalid_auth"}
+
+    except KustoServiceError as exp:
+        _LOGGER.error(exp)
+        return {"base": "cannot_connect"}
+
+    except Exception as exp:  # pylint: disable=broad-except
+        _LOGGER.error(exp)
+        return {"base": "unknown"}
+
+    return None
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Azure Data Explorer."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self._data: dict[str, Any] = {}
```

Comment:
you don't really need self._data, everything is done locally in the step_user function.